### PR TITLE
Add constraint keys

### PR DESCRIPTION
**Pull request by daniel-thom**
This PR adds constraint keys to replace use of symbols.

All tests pass except for two that try to read duals. Those are broken because dual symbols are not yet handled.

### CODE REVIEW

File: src/core/abstract_types.jl [hunk 1]
```diff
@@ -18,6 +18,21 @@ function encode_symbol(
     return Symbol("$(IS.strip_module_name(string(U)))_$(T_)" * meta_)
 end
 
+function encode_symbol(
+    ::Type{T},
+    ::Type{U},
+    ::Type{V},
+    meta::String = CONTAINER_KEY_EMPTY_META,
+) where {T <: PSY.Component, U, V}
+    # TODO DT: it would be better to use "__" in case we need to decode and some names have "_"
```
Comment (daniel-thom) [inline src/core/abstract_types.jl:27]: Do we want to do this? Or is it guaranteed that `_` won't exist in the fields?

File: src/core/feedforward.jl [hunk 2]
```diff
@@ -88,18 +88,18 @@ where r in range_data.
 * cons_name::Symbol : name of the constraint
 * var_key::VariableKey : the name of the continuous variable
 """
+# TODO DT: this function is not tested
```
Comment (daniel-thom) [inline src/core/feedforward.jl:91]: Should we create an issue for this?

File: src/core/feedforward.jl [hunk 3]
```diff
@@ -189,24 +191,37 @@ where r in range_data.
...
     set_name = [get_component_name(ci) for ci in constraint_infos]
     @assert axes[2] == time_steps
     container = add_param_container!(optimization_container, param_reference, set_name)
     multiplier = get_multiplier_array(container)
     param = get_parameter_array(container)
-    con_ub = add_cons_container!(optimization_container, ub_name, set_name, time_steps)
-    con_lb = add_cons_container!(optimization_container, lb_name, set_name, time_steps)
+    con_ub = add_cons_container!(
+        optimization_container,
+        constraint_type,
```
Comment (daniel-thom) [inline src/core/feedforward.jl:211]: Should this include the VariableType?

File: src/core/feedforward.jl [hunk 4]
```diff
@@ -366,14 +382,14 @@ function feedforward!(
...
+    for var_key in get_affected_variables(ff_model)
         semicontinuousrange_ff(
             optimization_container,
-            make_constraint_name(FEEDFORWARD_BIN, T),
+            FeedforwardBinConstraint,
+            T,
             constraint_infos,
             parameter_ref,
-            var_name,
+            get_entry_type(var_key),  # TODO DT: Jose, the old code was creating a new key; not sure why
```
Comment (daniel-thom) [inline src/core/feedforward.jl:392]: Need a reply here

File: src/core/feedforward.jl [hunk 5]
```diff
@@ -384,14 +400,14 @@ function feedforward!(
...
+    for var_key in get_affected_variables(ff_model)
         parameter_ref = UpdateRef{JuMP.VariableRef}(var_key)
         integral_limit_ff(
             optimization_container,
-            make_constraint_name(FEEDFORWARD_INTEGRAL_LIMIT, T),
+            FeedforwardIntegralLimitConstraint,
+            T,
             parameter_ref,
-            var_name,
+            get_entry_type(var_key),  # TODO DT: Jose, the old code was creating a new key; not sure why
```
Comment (daniel-thom) [inline src/core/feedforward.jl:410]: Need a reply here

File: src/core/update_initial_conditions.jl [hunk 6]
```diff
@@ -15,7 +15,8 @@ function calculate_ic_quantity(
     current_counter = time_cache[:count]
     last_status = time_cache[:status]
     var_status = isapprox(var_value, 0.0, atol = ABSOLUTE_TOLERANCE) ? 0.0 : 1.0
-    @debug last_status, var_status, abs(last_status - var_status)
+    @debug last_status, var_status, abs(last_status - var_status) _group =
```
Comment (daniel-thom) [inline src/core/update_initial_conditions.jl:18]: I added `_group` to some debug statements to make it easier to debug my work.

File: src/core/constraints.jl [hunk 7]
```diff
@@ -1,94 +1,106 @@
...
-function make_constraint_name(
+function ConstraintKey(
     ::Type{T},
-    ::Type{U},
-    ::Type{V},
-) where {T <: ConstraintType, U <: VariableType, V <: PSY.Device}
-    return encode_symbol(T, encode_symbol(V, U))
+    meta = CONTAINER_KEY_EMPTY_META,
+) where {T <: ConstraintType}
+    return ConstraintKey(T, VariableNotDefined, PSY.Component, meta)
```
Comment (daniel-thom) [inline src/core/constraints.jl:67]: Use of VariableNotDefined may be a point of disagreement.

File: src/devices_models/devices/common/duration_constraints.jl [hunk 8]
```diff
@@ -137,24 +149,26 @@ for i in the set of time steps.
 - : var_keys[2] : varstart
 - : var_keys[3] : varstop
 """
+# TODO DT: this function is not tested
```
Comment (daniel-thom) [inline src/devices_models/devices/common/duration_constraints.jl:152]: Should we create an issue for this?

File: src/devices_models/devices/common/energy_balance_constraint.jl [hunk 9]
```diff
@@ -163,14 +168,23 @@ function energy_balance!(
...
+        get_variable(optimization_container, var, inputs.component_type) for
+        var in inputs.pout_variable_types
+    ]
 
     constraint = add_cons_container!(
         optimization_container,
-        inputs.constraint_name,
+        inputs.constraint_type,
+        # TODO DT: is it correct to use this?
+        inputs.energy_variable,
```
Comment (daniel-thom) [inline src/devices_models/devices/common/energy_balance_constraint.jl:186]: Need a response here

File: src/devices_models/devices/common/energy_balance_constraint.jl [hunk 10]
```diff
@@ -259,7 +279,10 @@ function energy_balance_param!(
...
```
Comment (daniel-thom) [inline src/devices_models/devices/common/energy_balance_constraint.jl:284]: Need a response here

File: src/devices_models/devices/common/range_constraint.jl [hunk 11]
```diff
@@ -371,17 +471,31 @@ function reserve_power_ub!(
...
-    rev_dn_name = middle_rename(cons_name, PSI_NAME_DELIMITER, "dn")
+    var_in = get_variable(optimization_container, var_types[1], T)
+    var_out = get_variable(optimization_container, var_types[2], T)
     names = [get_component_name(x) for x in charging_range_data]
-    con_up = add_cons_container!(optimization_container, rev_up_name, names, time_steps)
-    con_dn = add_cons_container!(optimization_container, rev_dn_name, names, time_steps)
+    con_up = add_cons_container!(
+        optimization_container,
+        cons_type,
+        # TODO DT: we could pass a tuple of variables. Is there a chance for collisions?
```
Comment (daniel-thom) [inline src/devices_models/devices/common/range_constraint.jl:485]: Question for discussion

File: src/devices_models/devices/common/rateofchange_constraints.jl [hunk 12]
```diff
@@ -141,21 +156,36 @@ If t > 1:
...
+    varstart = get_variable(optimization_container, var_types[2], T)
+    varstop = get_variable(optimization_container, var_types[3], T)
 
     set_name = [get_component_name(r) for r in rate_data]
-    con_up = add_cons_container!(optimization_container, up_name, set_name, time_steps)
-    con_down = add_cons_container!(optimization_container, down_name, set_name, time_steps)
+    con_up = add_cons_container!(
+        optimization_container,
+        cons_type,
+        # TODO DT: should this use variable?
```
Comment (daniel-thom) [inline src/devices_models/devices/common/rateofchange_constraints.jl:174]: Need a response here

File: src/devices_models/devices/common/rateofchange_constraints.jl [hunk 13]
```diff
@@ -330,17 +374,25 @@ end
...
     service_name::AbstractString,
-)
+    ::Type{T},
+) where {T <: PSY.Component}
     time_steps = model_time_steps(optimization_container)
-    up_name = middle_rename(cons_name, PSI_NAME_DELIMITER, "up" * service_name)
 
-    variable = get_variable(optimization_container, var_key)
+    # TODO DT: is this change valid?
+    variable = get_variable(optimization_container, var_type, T, service_name)
```
Comment (daniel-thom) [inline src/devices_models/devices/common/rateofchange_constraints.jl:385]: Need a response here

File: src/devices_models/devices/thermal_generation.jl [hunk 14]
```diff
@@ -800,36 +788,39 @@ for t in time_limits[s+1]:T
...
+    varstop = get_variable(optimization_container, var_stop, T)
 
     names = [get_component_name(st) for st in startup_data]
 
     con = [
         add_cons_container!(
             optimization_container,
-            hot_name,
+            cons_type,
+            # TODO DT: should I use var_stop here?
```
Comment (daniel-thom) [inline src/devices_models/devices/thermal_generation.jl:809]: Need a response here

File: src/devices_models/devices/thermal_generation.jl [hunk 15]
```diff
@@ -876,23 +867,24 @@ Constructs contraints that restricts devices to one type of start at a time
...
-        get_variable(optimization_container, var_keys[2]),
-        get_variable(optimization_container, var_keys[3]),
+        get_variable(optimization_container, var_types[1], T),
+        get_variable(optimization_container, var_types[2], T),
+        get_variable(optimization_container, var_types[3], T),
     ]
 
     set_name = [get_component_name(d) for d in data]
-    con = add_cons_container!(optimization_container, cons_name, set_name, time_steps)
+    con = add_cons_container!(optimization_container, cons_type, T, set_name, time_steps)
```
Comment (daniel-thom) [inline src/devices_models/devices/thermal_generation.jl:887]: Should I use the type of var_start here?

File: src/network_models/copperplate_model.jl [hunk 16]
```diff
@@ -6,8 +6,12 @@ function copper_plate(
     time_steps = model_time_steps(optimization_container)
     expressions = get_expression(optimization_container, expression)
     remove_undef!(optimization_container.expressions[expression])
-    constraint =
-        add_cons_container!(optimization_container, :CopperPlateBalance, time_steps)
+    # TODO DT: is it correct to not specify a component type?
```
Comment (daniel-thom) [inline src/network_models/copperplate_model.jl:9]: Need a response here

File: src/core/abstract_types.jl (hunk 1 above)
Comment (jd-lara) [reply src/core/abstract_types.jl:27]: There shouldn't be any names with `_` anymore if I am correct.

File: src/core/feedforward.jl (hunk 2 above)
Comment (jd-lara) [reply src/core/feedforward.jl:91]: Yes, we should record this.

File: src/network_models/copperplate_model.jl (hunk 16 above)
Comment (jd-lara) [reply src/network_models/copperplate_model.jl:9]: Yeah, this doesn't have a component type right now. I am not sure what would be a correct way of specifying this.

File: src/devices_models/devices/common/energy_balance_constraint.jl [hunk 17]
```diff
@@ -1,36 +1,39 @@
 struct DeviceEnergyBalanceConstraintSpec
-    constraint_name::Symbol
-    energy_variable::VariableKey
+    constraint_type::ConstraintType
```
Comment (jd-lara) [inline src/devices_models/devices/common/energy_balance_constraint.jl:2]: This is a huge improvement, thanks for making these in here.

File: src/services_models/agc.jl [hunk 18]
```diff
@@ -146,8 +146,9 @@ end
 function absolute_value_lift(optimization_container::OptimizationContainer, areas)
     time_steps = model_time_steps(optimization_container)
     area_names = [PSY.get_name(a) for a in areas]
-    container_lb = add_cons_container!(optimization_container, :absolute_value_lb, area_names, time_steps)
-    container_ub = add_cons_container!(optimization_container, :absolute_value_ub, area_names, time_steps)
+    # TODO DT: correct component type and meta?
```
Comment (jd-lara) [inline src/services_models/agc.jl:149]: This looks correct

**Review by jd-lara — APPROVED** [review]
@daniel-thom thanks for the PR. This does exactly what I had in mind for the keys.

**Pull request merged**